### PR TITLE
Improve scavenger

### PR DESCRIPTION
**Pull request by BrunoScheufler**
## Description

scavenging is important to move abandoned in-progress items back to the "backlog" queue partition, so they get picked up by executors again. we observed some in-progress jobs not extending their lease for a prolonged duration, indicating that scavenging is not moving at the required pace, or at all.

- key queues should never be in there anymore: key queues are only used for tracking concurrency, NOT used as an execution entrypoint, so we don't care about key queues in the scavenger
- we ALWAYS want to add a pointer to the default fn partition to the scavenger queue, so that stalling jobs get picked up
- we should gracefully clean up leftover key queue data, and remove pointers to empty concurrency queues atomically. otherwise, we may accrue unnecessary pointers, which leaves the scavenger spinning without doing any work.

with these changes
- the scavenger queue (concurrency index) should gradually empty, only including pointers to function concurrency queues with actual in progress items
- key queues should be filtered out and never added again

context
- we have concurrency queues that contain stalled items. these should have been picked up by the scavenger
- the scavenger is logging key queues, which should no longer be in there
- the scavenger must clean up leftover items fast, so we cannot build up unnecessary cruft

## Type of change (choose one)
- [ ] Chore (refactors, upgrades, etc.)
- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] Security fix (non-breaking change that fixes a potential vulnerability)
- [ ] Docs
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

## Checklist
- [ ] I've linked any associated issues to this PR.
- [ ] I've tested my own changes.

*[Check our Pull Request Guidelines](https://github.com/inngest/inngest/blob/main/docs/PULL_REQUEST_GUIDELINES.md)*

### CODE REVIEW

File: pkg/execution/state/redis_state/queue.go [hunk 1]
```diff
@@ -1441,6 +1442,41 @@ func (q *queue) SetFunctionPaused(ctx context.Context, accountId uuid.UUID, fnID
 	return nil
 }
 
+// SetFunctionPaused sets the "Paused" flag (represented in JSON as "off") for the given
+// function ID's queue partition.
+// If a function is unpaused, we requeue the partition with a score of "now" to ensure that it is processed.
```
Comment (darwin67) [inline pkg/execution/state/redis_state/queue.go:1447]: what does this comment have to do with the function?

**Review by darwin67 — APPROVED** [review]

File: pkg/execution/state/redis_state/queue.go [hunk 2]
```diff
@@ -3114,28 +3166,48 @@ func (q *queue) Scavenge(ctx context.Context, limit int) (int, error) {
...
 			Key(queueKey).
 			Min("-inf").
 			Max(now).
 			Byscore().
-			Limit(0, 100).
+			Limit(0, ScavengeConcurrencyQueuePeekSize).
 			Build()
-		itemIDs, err := q.primaryQueueShard.RedisClient.unshardedRc.Do(ctx, cmd).AsStrSlice()
+		itemIDs, err := client.Do(ctx, cmd).AsStrSlice()
 		if err != nil && err != rueidis.Nil {
```
Comment (darwin67) [inline pkg/execution/state/redis_state/queue.go:3189]: so what happens here if it's an error but it's `rueidis.Nil`?
does it just move forward?

File: pkg/execution/state/redis_state/queue.go (hunk 1 above)
Comment (BrunoScheufler) [reply pkg/execution/state/redis_state/queue.go]: nothing, thanks for catching 🙃

File: pkg/execution/state/redis_state/queue.go (hunk 2 above)
Comment (BrunoScheufler) [reply pkg/execution/state/redis_state/queue.go:3189]: If it's rueidis nil. the ZSET is empty and will be handled by the next clause

**Pull request merged**